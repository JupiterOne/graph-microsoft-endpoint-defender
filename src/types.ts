import {
  IntegrationInstanceConfig,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The Azure application client ID used to identify the program as the
   * registered app. This will be the same for all tenants since this is a
   * multi-tenant application.
   */
  clientId: string;

  /**
   * The Azuser application client secret used to authenticate requests.
   */
  clientSecret: string;

  /**
   * The target directory/tenant ID, identified during the admin consent OAuth
   * flow.
   */
  tenant: string;

  /**
   * This is to differentiate between the azure api vs defender api **/
  isDefenderApi: boolean;
}

/**
 * An `IntegrationStepExecutionContext` typed for this integration's
 * `IntegrationInstanceConfig`.
 */
export type IntegrationStepContext =
  IntegrationStepExecutionContext<IntegrationConfig>;

export interface Machine {
  id: string;
  computerDnsName: string;
  firstSeen: string;
  lastSeen: string;
  classification: string;
  onboardingStatus: string;
  healthStatus: string;
  deviceValue: string;
  agentVersion: string;
  defenderAvStatus: string;
  riskScore: string;
  osPlatform: string;
  osVersion: any;
  osProcessor: string;
  version: string;
  lastIpAddress: string;
  lastExternalIpAddress: string;
  osBuild: number;
  rbacGroupId: number;
  rbacGroupName: string;
  exposureLevel: string;
  isAadJoined: boolean;
  aadDeviceId: any;
  machineTags: Array<string>;
  osArchitecture: string;
  managedBy: string;
  managedByStatus: string;
  ipAddresses: Array<IpAddress>;
  vmMetadata: VmMetadata;
}

export interface IpAddress {
  ipAddress: string;
  macAddress: string;
  type: string;
  operationalStatus: string;
}

export interface VmMetadata {
  vmId: string;
  cloudProvider: string;
  resourceId: string;
  subscriptionId: string;
}

export interface UserLogon {
  id: string;
  accountName: string;
  accountDomain: string;
  firstSeen: string;
  lastSeen: string;
  logonTypes: string;
  healthStatus: string;
  deviceValue: string;
  agentVersion: string;
  defenderAvStatus: string;
  isDomainAdmin: boolean;
  logOnMachinesCount: any;
  isOnlyNetworkUser: any;
  accountSid: any;
}

export interface Vulnerability {
  id: any;
  name: string;
  description: string;
  severity: string;
  cvssV3: any;
  exposedMachines: string;
  publishedOn: any;
  updatedOn: any;
  publicExploit: boolean;
  exploitVerified: boolean;
  exploitInKit: false;
  exploitTypes: Array<string>;
  exploitUris: Array<string>;
}

export interface Finding {
  id: any;
  name: string;
  description: string;
  severity: string;
  cvssV3: any;
  exposedMachines: string;
  publishedOn: any;
  updatedOn: any;
  publicExploit: boolean;
  exploitVerified: boolean;
  exploitInKit: false;
  exploitTypes: Array<string>;
  exploitUris: Array<string>;
}
