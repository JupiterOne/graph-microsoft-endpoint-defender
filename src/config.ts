import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationStepExecutionContext,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { GraphClient } from './ms-graph/client';
import { IntegrationInstanceConfigFieldMap } from '@jupiterone/integration-sdk-core';

/**
 * A type describing the configuration fields required to execute the
 * integration for a specific account in the data provider.
 *
 * When executing the integration in a development environment, these values may
 * be provided in a `.env` file with environment variables. For example:
 *
 * - `CLIENT_ID=123` becomes `instance.config.clientId = '123'`
 * - `CLIENT_SECRET=abc` becomes `instance.config.clientSecret = 'abc'`
 *
 * Environment variables are NOT used when the integration is executing in a
 * managed environment. For example, in JupiterOne, users configure
 * `instance.config` in a UI.
 */

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  clientId: {
    type: 'string',
  },
  clientSecret: {
    type: 'string',
    mask: true,
  },
  tenant: {
    type: 'string',
  },
  findingSeverity: {
    type: 'string',
  },
  findingsLimit: {
    type: 'string',
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The Azure application client ID used to identify the program as the
   * registered app. This will be the same for all tenants since this is a
   * multi-tenant application.
   */
  clientId: string;

  /**
   * The Azuser App Registration client secret used to authenticate requests.
   */
  clientSecret: string;

  /**
   * The target directory/tenant ID, identified during the admin consent OAuth
   * flow.
   */
  tenant: string;
}

/**
 * An `IntegrationStepExecutionContext` typed for this integration's
 * `IntegrationInstanceConfig`.
 */
export type IntegrationStepContext =
  IntegrationStepExecutionContext<IntegrationConfig>;

export function validateExecutionConfig(
  executionContext: IntegrationExecutionContext<IntegrationConfig>,
): void {
  const { clientId, clientSecret, tenant } = executionContext.instance.config;

  executionContext.logger.info(
    {
      clientId,
      tenantId: tenant,
    },
    'Configured to make Microsoft Graph API calls to tenantId acting as clientId',
  );

  if (!clientId || !clientSecret || !tenant) {
    throw new IntegrationValidationError(
      'Config requires all of {clientId, clientSecret, tenant}',
    );
  }
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  validateExecutionConfig(context);
  const apiClient = new GraphClient(context.logger, context.instance.config);
  await apiClient.verifyAuthentication();
}
