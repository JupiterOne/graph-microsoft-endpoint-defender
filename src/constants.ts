import {
  RelationshipClass,
  RelationshipDirection,
  StepEntityMetadata,
  StepMappedRelationshipMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const Steps: Record<
  | 'FETCH_ACCOUNT'
  | 'FETCH_MACHINES'
  | 'FETCH_LOGON_USERS'
  | 'FETCH_USERS'
  | 'FETCH_VULNERABILITIES'
  | 'VULNERABILITY_CVE_RELATIONSHIP'
  | 'FETCH_ENDPOINTS',
  { id: string; name: string }
> = {
  FETCH_ACCOUNT: { id: 'fetch-account', name: 'Fetch Account' },
  FETCH_MACHINES: { id: 'fetch-machines', name: 'Fetch Machines' },
  FETCH_LOGON_USERS: { id: 'fetch-logon-users', name: 'Fetch Logon Users' },
  FETCH_USERS: { id: 'fetch-users', name: 'Fetch Users' },
  FETCH_VULNERABILITIES: {
    id: 'fetch-vulnerabilities',
    name: 'Fetch Vulnerabilities',
  },
  VULNERABILITY_CVE_RELATIONSHIP: {
    id: 'build-vulnerability-cve-relationship',
    name: 'Build Vulnerability and CVE Relationship',
  },
  FETCH_ENDPOINTS: {
    id: 'fetch-endpoints',
    name: 'Fetch Endpoints',
  },
};

export const Entities: Record<
  'ACCOUNT' | 'MACHINE' | 'LOGON_USER' | 'USER' | 'VULNERABILITY' | 'ENDPOINT',
  StepEntityMetadata
> = {
  ACCOUNT: {
    resourceName: 'Account',
    _type: 'microsoft_defender_account',
    _class: ['Account'],
  },
  MACHINE: {
    resourceName: 'Machine',
    _type: 'microsoft_defender_machine',
    _class: ['HostAgent'],
  },
  LOGON_USER: {
    resourceName: 'Logon User',
    _type: 'microsoft_defender_logon_user',
    _class: ['User'],
  },
  USER: {
    resourceName: 'User',
    _type: 'microsoft_defender_user',
    _class: ['User'],
  },
  VULNERABILITY: {
    resourceName: 'Vulnerability',
    _type: 'microsoft_defender_vulnerability',
    _class: ['Finding'],
  },
  ENDPOINT: {
    resourceName: 'Device/Machine/Host',
    _type: 'user_endpoint',
    _class: ['Device'],
  },
};

export const Relationships: Record<
  | 'ACCOUNT_HAS_MACHINE'
  | 'ACCOUNT_HAS_USER'
  | 'MACHINE_HAS_LOGON_USER'
  | 'MACHINE_IDENTIFIED_VULNERABILITY'
  | 'MACHINE_MANAGES_ENDPOINT',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_MACHINE: {
    _type: 'microsoft_defender_account_has_machine',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.MACHINE._type,
  },
  ACCOUNT_HAS_USER: {
    _type: 'microsoft_defender_account_has_user',
    sourceType: Entities.ACCOUNT._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.USER._type,
  },
  MACHINE_HAS_LOGON_USER: {
    _type: 'microsoft_defender_machine_has_logon_user',
    sourceType: Entities.MACHINE._type,
    _class: RelationshipClass.HAS,
    targetType: Entities.LOGON_USER._type,
  },
  MACHINE_IDENTIFIED_VULNERABILITY: {
    _type: 'microsoft_defender_machine_identified_vulnerability',
    sourceType: Entities.MACHINE._type,
    _class: RelationshipClass.IDENTIFIED,
    targetType: Entities.VULNERABILITY._type,
  },
  MACHINE_MANAGES_ENDPOINT: {
    _type: 'microsoft_defender_machine_manages_user_endpoint',
    sourceType: Entities.MACHINE._type,
    _class: RelationshipClass.MANAGES,
    targetType: Entities.ENDPOINT._type,
  },
};

export const MappedRelationships: Record<
  'VULNERABILITY_IS_CVE',
  StepMappedRelationshipMetadata
> = {
  VULNERABILITY_IS_CVE: {
    _type: 'microsoft_defender_vulnerability_is_cve',
    sourceType: Entities.VULNERABILITY._type,
    _class: RelationshipClass.IS,
    targetType: 'cve',
    direction: RelationshipDirection.FORWARD,
  },
};

export const ACCOUNT_ENTITY_KEY = 'entity:account';
