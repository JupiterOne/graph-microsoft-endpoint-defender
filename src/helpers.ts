import { createIntegrationHelpers } from '@jupiterone/integration-sdk-core';
import { typeboxClassSchemaMap } from '@jupiterone/data-model';

export const { createEntityType, createEntityMetadata } =
  createIntegrationHelpers({
    integrationName: 'microsoft_defender',
    classSchemaMap: typeboxClassSchemaMap,
  });
