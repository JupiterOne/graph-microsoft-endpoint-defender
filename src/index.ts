import { IntegrationInvocationConfig } from '@jupiterone/integration-sdk-core';
import { activeDirectorySteps } from './steps/active-directory';
import {
  instanceConfigFields,
  IntegrationConfig,
  validateInvocation,
} from './config';
import { msDefenderSteps } from './steps/ms-defender';
import { ingestionConfig } from './ingestionConfig';

export const integrationSteps = [...activeDirectorySteps, ...msDefenderSteps];

export const invocationConfig: IntegrationInvocationConfig<IntegrationConfig> =
  {
    instanceConfigFields,
    validateInvocation,
    integrationSteps,
    ingestionConfig,
  };
