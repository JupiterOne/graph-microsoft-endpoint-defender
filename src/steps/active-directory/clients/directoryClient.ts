import { Finding, Machine, UserLogon } from '../../../types';
import { GraphClient } from '../../../ms-graph/client';

export enum MemberType {
  USER = '#microsoft.graph.user',
}

export class DirectoryGraphClient extends GraphClient {
  // https://docs.microsoft.com/en-us/microsoft-365/security/defender-endpoint/get-machines?view=o365-worldwide
  public async iterateMachines(
    callback: (user: Machine) => void | Promise<void>,
  ): Promise<void> {
    return this.iterateResources({
      resourceUrl: `${process.env.BASE_URL_API}/machines`,
      callback,
    });
  }

  // https://docs.microsoft.com/en-us/microsoft-365/security/defender-endpoint/get-machine-log-on-users?view=o365-worldwide
  public async iterateUsers(
    input: {
      machineId: string;
      /**
       * The property names for `$select` query param.
       */
      select?: string | string[];
    },
    callback: (user: UserLogon) => void | Promise<void>,
  ): Promise<void> {
    const $select = input.select
      ? Array.isArray(input.select)
        ? input.select.join(',')
        : input.select
      : undefined;
    return this.iterateResources({
      resourceUrl: `${process.env.BASE_URL_API}/machines/${input.machineId}/logonusers`,
      query: $select ? { $select } : undefined,
      callback,
    });
  }

  // https://docs.microsoft.com/en-us/microsoft-365/security/defender-endpoint/get-discovered-vulnerabilities?view=o365-worldwide
  public async iterateFindings(
    input: {
      machineId: string;
      /**
       * The property names for `$select` query param.
       */
      select?: string | string[];
    },
    callback: (user: Finding) => void | Promise<void>,
  ): Promise<void> {
    let filters: any = [];
    if (process.env.FINDING_SEVERITY) {
      filters.push(
        '$filter=severity+eq+' +
          process.env.FINDING_SEVERITY?.split(',')
            .map((z) => "'" + z + "'")
            .join('+or+severity+eq+'),
      );
    }
    if (process.env.FINDIGS_LIMIT) {
      filters.push('$top=' + process.env.FINDIGS_LIMIT);
    }

    if (filters && filters.length) {
      filters = filters.join('&');
    }

    const url = `${process.env.BASE_URL_API}/machines/${input.machineId}/vulnerabilities?${filters}`;
    return this.iterateResources({
      resourceUrl: url,
      callback,
    });
  }
}
