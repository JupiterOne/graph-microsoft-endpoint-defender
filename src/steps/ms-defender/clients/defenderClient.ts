import { Finding, Machine, UserLogon } from '../../../types';
import { GraphClient } from '../../../ms-graph/client';

export class DefenderClient extends GraphClient {
  BASE_URL_API: string = 'https://api.securitycenter.microsoft.com/api';
  constructor(logger, config) {
    super(logger, { ...config, isDefenderApi: true });
  }

  // https://docs.microsoft.com/en-us/microsoft-365/security/defender-endpoint/get-machines?view=o365-worldwide
  public async iterateMachines(
    callback: (user: Machine) => void | Promise<void>,
  ): Promise<void> {
    return this.iterateResources({
      resourceUrl: `${this.BASE_URL_API}/machines`,
      callback,
    });
  }

  // https://docs.microsoft.com/en-us/microsoft-365/security/defender-endpoint/get-machine-log-on-users?view=o365-worldwide
  public async iterateUsers(
    input: {
      machineId: string;
      /**
       * The property names for `$select` query param.
       */
      select?: string | string[];
    },
    callback: (user: UserLogon) => void | Promise<void>,
  ): Promise<void> {
    const $select = input.select
      ? Array.isArray(input.select)
        ? input.select.join(',')
        : input.select
      : undefined;
    return this.iterateResources({
      resourceUrl: `${this.BASE_URL_API}/machines/${input.machineId}/logonusers`,
      query: $select ? { $select } : undefined,
      callback,
    });
  }

  // https://docs.microsoft.com/en-us/microsoft-365/security/defender-endpoint/get-discovered-vulnerabilities?view=o365-worldwide
  public async iterateFindings(
    input: {
      machineId: string;
      /**
       * The property names for `$select` query param.
       */
      select?: string | string[];
    },
    callback: (user: Finding) => void | Promise<void>,
  ): Promise<void> {
    let filters: any = [];
    if (this.config.findingSeverity) {
      filters.push(
        '$filter=severity+eq+' +
          this.config.findingSeverity
            .split(',')
            .map((z) => "'" + z + "'")
            .join('+or+severity+eq+'),
      );
    }
    if (this.config.findingsLimit) {
      filters.push('$top=' + this.config.findingsLimit);
    }

    if (filters && filters.length) {
      filters = filters.join('&');
    }

    const url = `${this.BASE_URL_API}/machines/${input.machineId}/vulnerabilities?${filters}`;
    return this.iterateResources({
      resourceUrl: url,
      callback,
    });
  }
}
