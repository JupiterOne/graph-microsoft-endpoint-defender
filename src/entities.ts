import { SchemaType } from '@jupiterone/integration-sdk-core';
import { createEntityType, createEntityMetadata } from './helpers';

const StringNullUndefined = SchemaType.Optional(
  SchemaType.Union([SchemaType.String(), SchemaType.Null()]),
);

export const [ACCOUNT_ENTITY, assignAccount] = createEntityMetadata({
  resourceName: 'Account',
  _class: ['Account'],
  _type: createEntityType('account'),
  description: 'A Microsoft Defender Endpoint account',
  schema: SchemaType.Object({
    organizationName: StringNullUndefined,
    defaultDomain: SchemaType.Optional(SchemaType.String()),
    verifiedDomains: SchemaType.Optional(SchemaType.Array(SchemaType.String())),
  }),
});

export const [MACHINE_ENTITY, assignMachine] = createEntityMetadata({
  resourceName: 'Machine',
  _class: ['HostAgent'],
  _type: createEntityType('machine'),
  description: 'A Microsoft Defender Endpoint machine',
  schema: SchemaType.Object({
    firstSeenOn: SchemaType.Optional(SchemaType.Number()),
    agentVersion: SchemaType.String(),
    defenderAvStatus: SchemaType.String(),
    riskScore: SchemaType.String(),
    computerDnsName: SchemaType.String(),
    rbacGroupId: SchemaType.Number(),
    rbacGroupName: SchemaType.Union([SchemaType.String(), SchemaType.Null()]),
    machineTags: SchemaType.Array(SchemaType.String()),
    onboardingStatus: SchemaType.String(),
    managedBy: SchemaType.String(),
    managedByStatus: SchemaType.String(),
    ipAddress: SchemaType.Array(SchemaType.String()),
    macAddress: SchemaType.Array(SchemaType.String()),
    aadDeviceId: SchemaType.Union([SchemaType.String(), SchemaType.Null()]),
  }),
});

export const [LOGON_USER_ENTITY, assignLogonUser] = createEntityMetadata({
  resourceName: 'Logon User',
  _class: ['User'],
  _type: createEntityType('logon_user'),
  description: 'A Microsoft Defender Endpoint Logon User',
  schema: SchemaType.Object({
    domain: StringNullUndefined,
    firstSeenOn: SchemaType.Optional(SchemaType.Number()),
    lastSeenOn: SchemaType.Optional(SchemaType.Number()),
    logonTypes: SchemaType.String(),
  }),
});

export const [USER_ENTITY, assignUser] = createEntityMetadata({
  resourceName: 'User',
  _class: ['User'],
  _type: createEntityType('user'),
  description: 'A Microsoft Defender Endpoint user',
  schema: SchemaType.Object({
    businessPhones: SchemaType.Optional(SchemaType.Array(SchemaType.String())),
    givenName: StringNullUndefined,
    jobTitle: StringNullUndefined,
    mail: StringNullUndefined,
    mobilePhone: StringNullUndefined,
    officeLocation: StringNullUndefined,
    preferredLanguage: StringNullUndefined,
    surname: StringNullUndefined,
    userPrincipalName: StringNullUndefined,
  }),
});

export const [VULNERABILITY_ENTITY, assignVulnerability] = createEntityMetadata(
  {
    resourceName: 'Vulnerability',
    _class: ['Finding'],
    _type: createEntityType('vulnerability'),
    description: 'A Microsoft Defender Endpoint vulnerability',
    schema: SchemaType.Object({
      id: SchemaType.String(),
      publishedOn: SchemaType.Optional(SchemaType.Number()),
      exposedMachines: SchemaType.Number(),
      blocking: SchemaType.Boolean(),
    }),
  },
);

export const [ENDPOINT_ENTITY, assignEndpoint] = createEntityMetadata({
  resourceName: 'Device/Machine/Host',
  _class: ['Device'],
  _type: 'user_endpoint',
  description: 'A Microsoft Defender Endpoint entity',
  schema: SchemaType.Object({
    computerDnsName: SchemaType.String(),
    firstSeenOn: SchemaType.Optional(SchemaType.Number()),
    osPlatform: SchemaType.String(),
    osProcessor: SchemaType.Optional(SchemaType.String()),
    lastIpAddress: SchemaType.String(),
    lastExternalIpAddress: SchemaType.String(),
    agentVersion: SchemaType.String(),
    osBuild: SchemaType.Union([SchemaType.Number(), SchemaType.Null()]),
    healthStatus: SchemaType.String(),
    deviceValue: SchemaType.String(),
    rbacGroupId: SchemaType.Number(),
    rbacGroupName: SchemaType.Union([SchemaType.String(), SchemaType.Null()]),
    riskScore: SchemaType.String(),
    exposureLevel: SchemaType.String(),
    isAadJoined: SchemaType.Union([SchemaType.Boolean(), SchemaType.Null()]),
    aadDeviceId: SchemaType.Union([SchemaType.String(), SchemaType.Null()]),
    machineTags: SchemaType.Array(SchemaType.String()),
    defenderAvStatus: SchemaType.String(),
    onboardingStatus: SchemaType.String(),
    osArchitecture: SchemaType.String(),
    managedBy: SchemaType.String(),
    managedByStatus: SchemaType.String(),
    vmId: SchemaType.Optional(SchemaType.String()),
    cloudProvider: SchemaType.Optional(SchemaType.String()),
    resourceId: SchemaType.Optional(SchemaType.String()),
    subscriptionId: SchemaType.Optional(SchemaType.String()),
    ipAddresses: SchemaType.Array(SchemaType.String()),
    ipAddress: SchemaType.Array(SchemaType.String()),
    macAddress: SchemaType.Array(SchemaType.String()),
  }),
});
