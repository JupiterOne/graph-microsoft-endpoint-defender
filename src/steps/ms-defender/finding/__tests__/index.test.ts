import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import {
  entities,
  TargetEntities,
  fetchFindings,
  buildFindingIsCveRelationships,
} from '..';
import { IntegrationConfig } from '../../../../../src/config';
import {
  Recording,
  setupProjectRecording,
} from '../../../../../test/recording';
import { Finding } from '../../../../types';
import { DefenderClient } from '../../../ms-defender/clients/defenderClient';
import { createMockIntegrationLogger } from '@jupiterone/integration-sdk-testing';

const DEFAULT_CLIENT_ID = 'dummy-acme-client-id';
const DEFAULT_CLIENT_SECRET = 'dummy-acme-client-secret';
const DEFAULT_TENANT = 'dummy-tenant';

const config: IntegrationConfig = {
  clientId: 'a8626c1e-191d-4e8f-9cbc-a4a6e85104ac',
  clientSecret: 'L3X8Q~BO57QkoAsXMSkevrmVR2qiNh.qKEuzucAt',
  tenant: '5a721b05-53ed-4ed9-be02-aed28f11edbd',
};

const logger: any = createMockIntegrationLogger();

// See test/README.md for details
// let recording: Recording;
// afterEach(async () => {
//   recording ? await recording.stop() : null;
// });

describe('fetchFindings', () => {
  it('Should create a Vulnerability finding entity correctly when user has the correct permissions', async () => {
    // recording = setupProjectRecording({
    //   directory: __dirname,
    //   name: 'fetchFindings',
    // });

    const context = createMockStepExecutionContext({ instanceConfig: config });
    await fetchFindings(context);
    const client = new DefenderClient(logger, config);
    const vulnerabilityFindingEntities = context.jobState.collectedEntities;
    const resources: Finding[] = [];
    await client.iterateFindings(
      {
        machineId: 'ef9f015a-f7af-4764-aed1-88b91718eda6',
        select: ['id', 'displayName'],
      },
      (e) => {
        resources.push(e);
      },
    );
    expect(resources.length).toBe(0);
    expect(vulnerabilityFindingEntities.length).toBe(0);
    expect(vulnerabilityFindingEntities).toMatchGraphObjectSchema({
      _class: entities.FINDING._class,
    });
    expect(resources).toMatchSnapshot(resources);
  });
});

describe('CveRelationshipsFindings', () => {
  it('Should create a Vulnerability finding entity correctly when there is a CVE relationship', async () => {
    const context = createMockStepExecutionContext({ instanceConfig: config });
    await buildFindingIsCveRelationships(context);

    const cveFindingEntities = context.jobState.collectedEntities;

    expect(cveFindingEntities.length).toBe(0);
    expect(cveFindingEntities).toMatchGraphObjectSchema({
      _class: TargetEntities.CVE._class,
    });
    expect(cveFindingEntities).toMatchSnapshot('cveFindingEntitiesSuccessful');
  });
});
