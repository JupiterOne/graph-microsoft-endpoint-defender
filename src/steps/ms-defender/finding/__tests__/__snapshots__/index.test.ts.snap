// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CveRelationshipsFindings Should create a Vulnerability finding entity correctly when there is a CVE relationship: cveFindingEntitiesSuccessful 1`] = `Array []`;

exports[`fetchFindings Should create a Vulnerability finding entity correctly when user has the correct permissions: vulnerabilityFindingEntitiesSuccessful 1`] = `Array []`;
